Add tagHelperPrefix directive.

NTaylorMullen opened the pull request:
Took the approach of attaching the prefix to `TagHelperDescriptor`s to enable tooling to separate `Prefix`s from `TagName`s in the document. Ran design by @rynowak.

*Change notes:*

- Updated TagHelperDescriptor to have a Prefix property. This enables new tooling scenarios such as refactoring prefixes or even giving them their own classification.
- Added invalid prefix cases '@', '!' and whitespace.
- Added TagHelperPrefix chunks, codegenerator, parsing logic to flow the directive through the Razor pipeline.
 Added parsing, sub-rewriting, rewriting and code generation tests for runtime/designtime.
- Fixed existing tests to utilize new class structures.
- Expanded block types to enable easy testing of prefixed TagHelpers.

#309

/cc @dougbu

dougbu commented:
> Added invalid prefix cases '@', '!' and whitespace

No to a blacklist.  Instead confirm a white list, specifically the prefix must contain only allowed HTML tag name characters, including `'-'` and `':'`.  Or just reuse whatever Razor does to confirm tag name composition for now and file a bug to confirm those characters correctly.

dougbu commented:
> Updated TagHelperDescriptor to have a Prefix property. This enables new tooling scenarios such as refactoring prefixes or even giving them their own classification.

This is not a feature at this time.  Why build in support for a _future_ enhancement when the current feature requires only that `TagHelperDescriptor` contains the correctly-prefixed tag name?

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs:
```diff
@@ -32,19 +32,25 @@ public static class TagHelperDescriptorFactory
+                    tagName,
+                    typeName,
+                    assemblyName,
+                    attributeDescriptors));
```
so much simpler and less-invasive if you create an unchanged `TagHelperDescriptor` with `prefix + tagName` for the `tagName` argument

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs:
```diff
@@ -89,7 +97,7 @@ public IEnumerable<TagHelperDescriptor> Resolve([NotNull] TagHelperDescriptorRes
                         Resources.FormatTagHelperDescriptorResolver_EncounteredUnexpectedError(
                             "@" + directiveName,
-                            directiveDescriptor.LookupText,
+                            directiveDescriptor.DirectiveText,
```
why did this line change?

NTaylorMullen replied:
`directiveDescriptor` is now used for `@addTagHelper`, `@removeTagHelper`, `@tagHelperPrefix` so `LookupText` didn't make sense.

dougbu replied:
ask I ask later -- why reuse the same `TagHelperDirectiveDescriptor` type?  you're mixing concepts and then the code that consumes it has to split things again.

NTaylorMullen replied:
Talked in person.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs:
```diff
@@ -32,19 +32,25 @@ public static class TagHelperDescriptorFactory
         /// Creates a <see cref="TagHelperDescriptor"/> from the given <paramref name="type"/>.
         /// </summary>
         /// <param name="assemblyName">The assembly name that contains <paramref name="type"/>.</param>
+        /// <param name="prefix">The prefix to the created <see cref="TagHelperDescriptor"/>s.</param>
```
nit: "to the" -> "for tag names in" throughout PR

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs:
```diff
@@ -104,26 +112,85 @@ public IEnumerable<TagHelperDescriptor> Resolve([NotNull] TagHelperDescriptorRes
+            var prefixDirectiveDescriptors = context.DirectiveDescriptors.Where(
+                descriptor => descriptor.DirectiveType == TagHelperDirectiveType.TagHelperPrefix);
+
+            if (prefixDirectiveDescriptors.Count() > 1)
```
this extra condition is unnecessary since `Skip(1)` will simply return an empty enumeration if `prefixDirectiveDescriptors` is empty or contains one item.

NTaylorMullen replied:
:+1:

dougbu commented on src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpDesignTimeHelpersVisitor.cs:
```diff
@@ -81,13 +86,9 @@ private void VisitAddOrRemoveTagHelperChunk(string lookupText, Chunk chunk)
-                    CultureInfo.InvariantCulture,
-                    "\"{0}\"", lookupText),
-                chunk);
+            _csharpCodeVisitor.CreateExpressionCodeMapping($"\"{text}\"", chunk);
```
this changes the `IFormatProvider` without reason.  leave as-is.

dougbu commented on src/Microsoft.AspNet.Razor/Parser/CSharpCodeParser.Directives.cs:
```diff
@@ -27,14 +28,27 @@ private void SetupDirectives()
 
+        protected virtual void TagHelperPrefixDirective()
+        {
+            TagHelperDirective(
```
why reuse this method for an additional, fairly-different purpose?  a new method for the prefix directive would be simpler and you would only need to change the existing method's name

NTaylorMullen replied:
:-1: The format of all tag helper directives is the same, from a parsing perspective they're nearly identical, making a whole other method where the largest change is the CodeGenerator isn't worth it.

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDirectiveDescriptor.cs:
```diff
@@ -11,23 +11,24 @@ namespace Microsoft.AspNet.Razor.TagHelpers
     {
         // Internal for testing purposes.
-        internal TagHelperDirectiveDescriptor(string lookupText,
+        internal TagHelperDirectiveDescriptor(string directiveText,
```
existing class is well-named for existing purpose.  why extend it to support this new, quite different directive?

NTaylorMullen replied:
It's a `TagHelper` related directive, all of which have the same basic format.

dougbu replied:
admit I may be incorrect about `TagHelperDirective`.  but at this point you're dealing with separate concepts that are handled on distinct code paths in the consuming code.  no longer dealing with format at all.

NTaylorMullen replied:
Sure the end purpose of the tag helper directive may be different but their core concepts are all the same: affect the creation of `TagHelperDescriptor`s  Their core concept is handled by the `TagHelperDescriptorResolver` who only understands provided `TagHelperDirectiveDescriptor`s, no reason to build a separate code path outside of this core `TagHelperDescriptor` building mechanism

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptor.cs:
```diff
@@ -28,6 +46,7 @@ public class TagHelperDescriptor
         /// Instantiates a new instance of the <see cref="TagHelperDescriptor"/> class with the given
         /// <paramref name="attributes"/>.
         /// </summary>
+        /// <param name="prefix">The prefix of the HTML element name the tag helper targets.</param>
```
prefix is a global setting, nothing near the level of individual tag helpers or their descriptions.  do not change this class.

dougbu replied:
thinking about this a bit and talking a couple of things through w/ @rynowak, I realize you've got two uses of these descriptors.  on one hand you have descriptor resolution, where the match between add/removeTagHelper directives and discovered tag helpers should use the original (from the tag helper) tag name.  on the other hand, the rewriter needs to match Razor tags against the full name of a tag helper.  note the prefix serves neither purpose and causes repeated string concatenations.

I suggest:
- add the `Prefix` to the `TagHelperDescriptorResolutionContext` (or another high-level context) so tooling can use it (if necessary)
- add a `string ShortName { get; }` property to `TagHelperDescriptor` and change the tag helper filtering code to use that instead.  seems like the less-used `TagName` requirement.  but flip it around to be `TagName` and `FullName` if that minimizes the churn.

NTaylorMullen replied:
First bullet point: `TagHelperDescriptor`'s building process logic exists in the `TagHelperDescriptorResolver`; therefore I feel that anything that affects them should also exist in the `TagHelperDescriptorResolver`, not to mention they all hold the same basic string format/directive structure. 

For your second bullet point I don't see the value in abstracting the `Prefix` into a combination of the element name + prefix from a `TagHelperDescriptor`s perspective. The `TagHelperDescriptor` should be extremely dumb/purely informational, consumers should deal with understanding what it applies to in the document.

dougbu replied:
the information to put in the extremely dumb `TagHelperDescriptor` is what the code requires.  everything is doing matches against either the short or the long name.  so provide those.  don't force repeated concatenations or surface a "prefix" concept when the code just wants a single string to match.

NTaylorMullen replied:
I'd much rather not add complexity to the `TagHelperDescriptor`, it should represent everything that makes up a valid `TagHelper`, in this case the HTMLs `TagName` (what should eventually hit the page) and its optional `Prefix`. Not to mention the editor uses these `TagHelperDescriptor`s and will also use the `Prefix` property directly.

dougbu commented:
:watch:

NTaylorMullen commented:
@dougbu sadly Razor doesn't use any complex logic to validate a tag name which is why I used the blacklist approach.

NTaylorMullen commented:
Updated.

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs:
```diff
@@ -51,7 +51,12 @@ public IEnumerable<TagHelperDescriptor> Resolve([NotNull] TagHelperDescriptorRes
         {
```
also update `_directiveNames` at the top of this class

NTaylorMullen replied:
:+1:

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs:
```diff
@@ -124,6 +133,79 @@ public IEnumerable<TagHelperDescriptor> Resolve([NotNull] TagHelperDescriptorRes
+                        SyntaxConstants.CSharp.TagHelperPrefixKeyword));
+            }
+
+            var prefixDirective = prefixDirectiveDescriptors.FirstOrDefault();
```
now see you could do a single loop through `prefixDirectiveDescriptors`.  might be more clear e.g.
``` c#
TagHelperDirective prefixDirective = null;
foreach (var directive in prefixDirectiveDescriptors)
{
  if (prefixDirective == null)
  {
    prefixDirective = directive;
  }
  else
  {
    context.ErrorSink.OnError(...);
  }
}
```

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs:
```diff
@@ -124,6 +133,79 @@ public IEnumerable<TagHelperDescriptor> Resolve([NotNull] TagHelperDescriptorRes
+                descriptor => descriptor.DirectiveType == TagHelperDirectiveType.TagHelperPrefix);
+
+            // For each invalid @tagHelperPrefix we need to create an error.
+            foreach (var erroredContext in prefixDirectiveDescriptors.Skip(1))
```
`erroredContext` -> `directive` ("errored" isn't a word and the variable doesn't hold a context)

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs:
```diff
@@ -124,6 +133,79 @@ public IEnumerable<TagHelperDescriptor> Resolve([NotNull] TagHelperDescriptorRes
+
+            if (prefix == null || !EnsureValidPrefix(prefix, prefixDirective.Location, context.ErrorSink))
+            {
+                prefix = string.Empty;
```
why not return `null`?

NTaylorMullen replied:
No reason anymore, :+1:

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs:
```diff
@@ -124,6 +133,79 @@ public IEnumerable<TagHelperDescriptor> Resolve([NotNull] TagHelperDescriptorRes
+                    // Reserved Razor transition symbol
+                    character == '@' ||
+                    // Reserved TagHelper opt-out character.
+                    character == '!')
```
from `HtmlSymbolType` I'd expect this blacklist (generally poor idea despite the existing mess w.r.t. character exclusion but :ok: for now) to be much longer.  should at least exclude everything that tokenizer will ensure is not found in a `HtmlSymbolType.Text` token.

dougbu commented on src/Microsoft.AspNet.Razor/Generator/TagHelpers/TagHelperPrefixDirectiveCodeGenerator.cs:
```diff
@@ -0,0 +1,44 @@
+namespace Microsoft.AspNet.Razor.Generator
+{
+    /// <summary>
+    /// A <see cref="SpanCodeGenerator"/> responsible for generating <see cref="Compiler.TagHelperPrefixDirectiveChunk"/>s.
```
long line

dougbu commented on src/Microsoft.AspNet.Razor/Generator/Compiler/CodeTree/Chunks/TagHelpers/TagHelperPrefixDirectiveChunk.cs:
```diff
@@ -0,0 +1,16 @@
+    public class TagHelperPrefixDirectiveChunk : Chunk
+    {
+        /// <summary>
+        /// Text used as the prefix for <see cref="TagHelpers.TagHelperDescriptor"/>s.
```
nit: storing a value in a descriptor isn't a "use". suggest `Text used as a required prefix when matching HTML start and end tags in the Razor source to available tag helpers.`  (might work without "in the Razor source")

similar comments about the descriptor rather than actual use appear elsewhere in this PR.

dougbu commented on src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs:
```diff
@@ -28,7 +28,11 @@ public static class TagHelperBlockRewriter
-            return new TagHelperBlockBuilder(tagName, selfClosing, start, attributes, descriptors);
+            // All descriptors for an element should share a common prefix, grab the first.
+            var prefix = descriptors.First().Prefix;
+            var tagHelperTagName = tagName.Substring(prefix.Length);
```
why does this code exist?  the descriptor should contain all `string`s of interest; no `Substring` required

NTaylorMullen replied:
We want the `TagName` to be the exact case that it is in the document, a `TagHelperDescriptor` is case insensitive where this block element is here to represent what the user typed.

dougbu replied:
to confirm: when all associated tag helpers no-op, the prefix will _not_ appear in the generated HTML.  if so, fine, that sounds consistent with the opt-out (`!`) handling.

but the prefix isn't used later.  need only the short name when generating the `TagHelperScopeManager.Begin()` call.  so passing it around together with descriptors containing exactly the same `string` is not worthwhile.

what may be useful is calculating the global prefix length once per parse, say at the top of this class.  then to get the short name from the full matched text, only need the `Substring` call.

NTaylorMullen replied:
`TagHelperBlockRewriter` is a static class, will not pass the prefix onto the `TagHelperBlock`/`TagHelperBlockBuilder` but will need to calculate the prefix length for each rewrite.

NTaylorMullen replied:
Actually scratch that, need this for testing scenarios and also in the `TagHelperParseTreeRewriter` matching scenario, we can't just pull from the `TagHelperDescriptor`s because their tagname may be `*`, just realized the `*` case is a hole I need to fix.

dougbu commented on src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlock.cs:
```diff
@@ -42,6 +43,11 @@ public TagHelperBlock(TagHelperBlockBuilder source)
         /// <summary>
+        /// The tag helper prefix.
+        /// </summary>
+        public string Prefix { get; }
```
you have the `Descriptors`, why surface the `Prefix` again?

dougbu commented on src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockBuilder.cs:
```diff
@@ -75,6 +83,11 @@ public TagHelperBlockBuilder(TagHelperBlock original)
         /// <summary>
+        /// The tag helper prefix.
+        /// </summary>
+        public string Prefix { get; set; }
```
have the descriptors; why is this exposed?

dougbu commented on src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs:
```diff
@@ -141,11 +141,15 @@ private bool TryRewriteTagHelper(Block tagBlock, RewritingContext context)
-                if (tagNameScope.Equals(tagName, StringComparison.OrdinalIgnoreCase))
+                if (parentTagHelper != null &&
+                    string.Equals(
+                        parentTagHelper.Prefix + parentTagHelper.TagName,
```
do not do another concatenation.  something is wrong with the information available to this method.  probably best to simply have `parentTagHelper.TagName` contain the full name that was matched for the start tag.

NTaylorMullen replied:
Have to based on our convo here: https://github.com/aspnet/Razor/pull/313#discussion_r25537282

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptor.cs:
```diff
@@ -35,23 +36,64 @@ public class TagHelperDescriptor
-                                   [NotNull] string typeName,
-                                   [NotNull] string assemblyName,
-                                   [NotNull] IEnumerable<TagHelperAttributeDescriptor> attributes)
+        public TagHelperDescriptor(
```
what call sites remain for this legacy constructor?

NTaylorMullen replied:
This is the ctor used by the `TagHelperDescriptorFactory`

dougbu replied:
:ok:

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptorComparer.cs:
```diff
@@ -29,12 +29,14 @@ public class TagHelperDescriptorComparer : IEqualityComparer<TagHelperDescriptor
         /// <remarks>
         /// Determines equality based on <see cref="TagHelperDescriptor.TypeName"/>,
-        /// <see cref="TagHelperDescriptor.AssemblyName"/> and <see cref="TagHelperDescriptor.TagName"/>.
+        /// <see cref="TagHelperDescriptor.AssemblyName"/> <see cref="TagHelperDescriptor.TagName"/> and
```
nit: comma after `<see cref="TagHelperDescriptor.AssemblyName"/>`

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptor.cs:
```diff
@@ -35,23 +36,64 @@ public class TagHelperDescriptor
         /// <summary>
+        /// The prefix of the HTML element name the tag helper targets.
+        /// </summary>
+        public string Prefix { get; private set; }
```
why does this have a setter?

NTaylorMullen replied:
This class was built so it could eventually be serialized.

dougbu replied:
I asked due to the visible inconsistency between `Prefix` and `FullTagName`.  Guess it's `FullTagName` that needs to change.

NTaylorMullen replied:
FullTagName is a meta property, that's set via the ctor, it doesn't need to have a setter.

dougbu replied:
constructors aren't run during deserialization

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptor.cs:
```diff
@@ -35,23 +36,64 @@ public class TagHelperDescriptor
+        /// <summary>
         /// The tag name that the tag helper should target.
         /// </summary>
         public string TagName { get; private set; }
```
while here only: why does this have a setter?

NTaylorMullen replied:
Same answer as above

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptor.cs:
```diff
@@ -35,23 +36,64 @@ public class TagHelperDescriptor
+        /// The full tag name that is required for the tag helper to target an HTML element.
+        /// </summary>
+        /// <remarks>This is equivalent to <see cref="Prefix"/> and <see cref="TagName"/> concatenated.</remarks>
+        public string FullTagName { get; }
```
this property is set but unused otherwise.  should be used (exclusively) when matching tag helper names from the Razor source.

NTaylorMullen replied:
:+1:

NTaylorMullen replied:
This actually can't happen due to catch-all tag helpers

dougbu replied:
need `private set;` or value will always be `null` after deserialization

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptorProvider.cs:
```diff
@@ -76,7 +78,9 @@ private void Register(TagHelperDescriptor descriptor)
-                _registrations[descriptor.TagName] = descriptorSet;
+
+                // We want to register TagHelperDescriptors by their prefix and tag name for easy lookup.
+                _registrations[descriptor.Prefix + descriptor.TagName] = descriptorSet;
```
use `descriptor.FullTagName` here too

NTaylorMullen replied:
This is the answer to this that I agreed on: https://github.com/aspnet/Razor/pull/313#discussion_r25491125

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorResolverTest.cs:
```diff
@@ -41,6 +42,465 @@ private static TagHelperDescriptor Valid_InheritedTagHelperDescriptor
+                {
+                    {
+                        defaultAssemblyLookups,
+                        new []
```
tiny tiny nit: no space between `new` and `[]`

dougbu commented:
:watch:

NTaylorMullen commented:
Updated.

dougbu commented on src/Microsoft.AspNet.Razor/Generator/Compiler/CodeTree/Chunks/TagHelpers/TagHelperPrefixDirectiveChunk.cs:
```diff
@@ -0,0 +1,17 @@
+namespace Microsoft.AspNet.Razor.Generator.Compiler
+{
+    /// <summary>
+    /// A <see cref="Chunk"/> for the @tagHelperPrefix directive.
```
nit: <c>@tagHelperPrefix</c>

dougbu commented on src/Microsoft.AspNet.Razor/Generator/TagHelpers/TagHelperPrefixDirectiveCodeGenerator.cs:
```diff
@@ -0,0 +1,45 @@
+        /// </summary>
+        /// <param name="prefix">
+        /// Text used as <see cref="TagHelpers.TagHelperDescriptor"/>'s 
+        /// <see cref="TagHelpers.TagHelperDescriptor.Prefix"/>.
```
use your "Text used as a required prefix when matching HTML ..." wording in all relevant doc comments

dougbu commented on src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs:
```diff
@@ -124,6 +136,85 @@ public IEnumerable<TagHelperDescriptor> Resolve([NotNull] TagHelperDescriptorRes
+                        Resources.FormatTagHelperDescriptorResolver_InvalidTagHelperDirective(
+                            SyntaxConstants.CSharp.TagHelperPrefixKeyword));
+                }
+
```
nit: extra blank line

dougbu commented on src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlock.cs:
```diff
@@ -42,6 +43,12 @@ public TagHelperBlock(TagHelperBlockBuilder source)
+        /// Text used as a required prefix when matching HTML start and end tags in the Razor source to available 
+        /// tag helpers.
+        /// </summary>
+        public string Prefix { get; }
```
this property should not exist.  it's currently used only when creating a `TagHelperBlockBuilder` from a `TagHelperBlock original`.  that overload is unused.

NTaylorMullen replied:
For the `TagHelperBlock` this property is used for testing, I could mark it internal.

dougbu replied:
should not have extra members (other than sometimes extra constructor overloads) in product code just for testing.  why test anything related to the prefix?  the important things are matching tags with descriptors and start tags with end tags.

dougbu commented on src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockBuilder.cs:
```diff
@@ -75,6 +83,12 @@ public TagHelperBlockBuilder(TagHelperBlock original)
+        /// Text used as a required prefix when matching HTML start and end tags in the Razor source to available 
+        /// tag helpers.
+        /// </summary>
+        public string Prefix { get; set; }
```
this class should have a single purpose -- gathering the information necessary to create a `TagHelperBlock`.  this property exists because it's also used on a stack in `TagHelperParseTreeRewriter`.  if you must use it that way, save the `FullTagName` you actually need in `TagHelperParseTreeRewriter.TryRewriteTagHelper()` to match an end tag to its start tag.

btw the current code saves the descriptor's prefix and so matches `<SomeTag>` with `</SomeTag>` if the configured prefix is "some" only because you're doing a case-insensitive match.  will be much more efficient if you save the exact text from the Razor source because the user likely used the same case.  (side note: I'm also not sure if HTML case-insensitivity extends to matching start and end tags with different cases.)

NTaylorMullen replied:
Will make a `FullTagName` property, remove this and the `FullTagName` property will calculate it on the fly by looking at the descriptors. This way it makes the builder accurate depending on its descriptors and reduces the amount of work we do outside of the builder.

dougbu replied:
sounds right but please clarify "calculate it": what needs to be calculated?  since a `TagHelperBlockBuilder` is created when full name (from the Razor source) is known, I hope you don't mean `FullTagName`.

dougbu commented on src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs:
```diff
@@ -28,7 +28,11 @@ public static class TagHelperBlockRewriter
+            var prefix = descriptors.First().Prefix;
+            var tagHelperTagName = tagName.Substring(prefix.Length);
+
+            return new TagHelperBlockBuilder(prefix, tagHelperTagName, selfClosing, start, attributes, descriptors);
```
as mentioned in `TagHelperBlockBuilder` either don't pass `prefix` or pass `tagName` (to become `FullTagName`) instead

dougbu commented on src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs:
```diff
@@ -28,7 +28,11 @@ public static class TagHelperBlockRewriter
 
-            return new TagHelperBlockBuilder(tagName, selfClosing, start, attributes, descriptors);
+            // All descriptors for an element should share a common prefix, grab the first.
+            var prefix = descriptors.First().Prefix;
```
neither `prefix` nor `prefix.Length` change for the lifetime of a `TagHelperParseTreeRewriter` instance.  calculate length once in that class's constructor and pass it into this method.

dougbu commented on src/Microsoft.AspNet.Razor/ParserResults.cs:
```diff
@@ -81,5 +82,10 @@ public class ParserResults
         public IEnumerable<TagHelperDescriptor> TagHelperDescriptors { get; }
+
+        /// <summary>
+        /// The prefix required to use tag helpers.
```
"Text used as a required prefix when matching HTML ..." wording makes sense here too.

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDirectiveType.cs:
```diff
@@ -16,6 +16,11 @@ public enum TagHelperDirectiveType
+        RemoveTagHelper,
+
+        /// <summary>
+        /// A @tagHelperPrefix directive.
```
nit (because comment is currently consistent with that of other values): `<c>@tagHelperPrefix</c>`

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorResolverTest.cs:
```diff
@@ -41,6 +42,465 @@ private static TagHelperDescriptor Valid_InheritedTagHelperDescriptor
+                        new[]
+                        {
+                            new TagHelperDirectiveDescriptor(
+                                "th:",
```
test other valid prefixes such as "th-" or just "th".  current coverage makes it seem like the colon is required.

NTaylorMullen replied:
done

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptorProvider.cs:
```diff
@@ -72,11 +80,21 @@ private void Register(TagHelperDescriptor descriptor)
+            var registrationKey = 
+                string.Equals(descriptor.TagName, CatchAllDescriptorTarget, StringComparison.Ordinal) ? 
+                CatchAllDescriptorTarget : 
+                descriptor.FullTagName;
```
nit: trailing whitespace.  suggest installing the trailing whitespace visualizer (thx @madskristensen !)

dougbu commented on src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockBuilder.cs:
```diff
@@ -75,6 +83,12 @@ public TagHelperBlockBuilder(TagHelperBlock original)
         }
 
         /// <summary>
+        /// Text used as a required prefix when matching HTML start and end tags in the Razor source to available 
```
nit: trailing whitespace

dougbu commented on test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorResolverTest.cs:
```diff
@@ -41,6 +42,465 @@ private static TagHelperDescriptor Valid_InheritedTagHelperDescriptor
+            }
+        }
+
+        public static TheoryData ResolveDirectiveDescriptorsTagHelperPrefixData
```
test a `TagHelperDirective` of type `TagHelperDirectiveType.TagHelperPrefix` with an empty prefix.  is that allowed or not?  can't tell from the tests in this class.

NTaylorMullen replied:
done

dougbu commented on test/Microsoft.AspNet.Razor.Test/Generator/CSharpTagHelperRenderingTest.cs:
```diff
@@ -59,50 +30,64 @@ public static TheoryData TagHelperDescriptorFlowTestData
+                        "BasicTagHelpers.Prefixed",
+                        "BasicTagHelpers.Prefixed",
+                        BuildPAndInputTagHelperDescriptors(prefix: "th:"),
+                        BuildPAndInputTagHelperDescriptors(prefix: "th:"),
```
- vary the valid prefixes in these more end-to-end tests as well
- nit: create a property containing the descriptors with this prefix.  (could do that using a deep copy of `DefaultPAndInputTagHelperDescriptors` instead of churning that code.)

NTaylorMullen replied:
Will change prefix here, going to keep the BuildDescriptor bits

dougbu commented on test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpDirectivesTest.cs:
```diff
@@ -13,6 +13,88 @@ namespace Microsoft.AspNet.Razor.Test.Parser.CSharp
     public class CSharpDirectivesTest : CsHtmlCodeParserTestBase
     {
         [Fact]
+        public void TagHelperPrefixDirective_Succeeds()
```
also test `"@tagHelperPrefix \"\""`

dougbu commented on test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperDescriptorProviderTest.cs:
```diff
@@ -10,7 +10,78 @@ namespace Microsoft.AspNet.Razor.Test.TagHelpers
     {
         [Fact]
-        public void TagHelperDescriptorProvider_GetTagHelpersReturnsNothingForUnregisteredTags()
+        public void GetTagHelpers_OnlyUnderstandsSinglePrefix()
```
test what happens when a catch-all descriptor has prefix "th" and `// Act` calls `provider.GetTagHelpers("th")`.  suspect for example `<th/>` will lead to an odd error or generation of a self-closing tag with no name.

NTaylorMullen replied:
:+1:, adding case in the `TagHelperDescriptorProvider`.

dougbu commented on test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs:
```diff
@@ -19,6 +19,142 @@ namespace Microsoft.AspNet.Razor.Test.TagHelpers
 {
     public class TagHelperParseTreeRewriterTest : CsHtmlMarkupParserTestBase
     {
+        public static TheoryData PrefixedTagHelperBoundData
```
- mix up prefixes
- test `<myth />` when prefix is "myth"

NTaylorMullen replied:
done

dougbu commented:
:watch:

NTaylorMullen commented:
Updated. Moved all prefix/Full tag name logic for the `TagHelperBlock`/`TagHelperBlockBuilder` into the `TagHelperCodeGenerator`, this way the `SyntaxTreeNode`s only ever represent what is in the document (the full tag name).

dougbu commented on src/Microsoft.AspNet.Razor/Generator/TagHelperCodeGenerator.cs:
```diff
@@ -75,9 +75,11 @@ public override void GenerateStartBlockCode(Block target, CodeGeneratorContext c
                 codeGenerator.Context.CodeTreeBuilder = new CodeTreeBuilder();
             }
 
+            var unprefixedTagName = tagHelperBlock.TagName.Substring(_tagHelperDescriptors.First().Prefix.Length);
```
nit: though this is done only once per `TagHelperCodeGenerator`, might be slightly cleaner to set a `readonly int _prefixLength` field in the constructor

dougbu commented on src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs:
```diff
@@ -141,11 +141,15 @@ private bool TryRewriteTagHelper(Block tagBlock, RewritingContext context)
                 // We're in an end tag helper block.
 
-                var tagNameScope = _tagStack.Count > 0 ? _tagStack.Peek().TagName : string.Empty;
+                var parentTagHelper = _tagStack.Count > 0 ? _tagStack.Peek() : null;
```
don't need to change this file.  back to using just the `TagName` value (or `null`).

dougbu commented on src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptorProvider.cs:
```diff
@@ -43,10 +44,18 @@ public IEnumerable<TagHelperDescriptor> GetTagHelpers(string tagName)
-            // Ensure there's an ISet to use.
+            if (!string.IsNullOrEmpty(_tagHelperPrefix) &&
+                (!tagName.StartsWith(_tagHelperPrefix, StringComparison.OrdinalIgnoreCase) ||
+                string.Equals(tagName, _tagHelperPrefix, StringComparison.OrdinalIgnoreCase)))
```
nit: easier to read and avoids repeated comparisons of the same characters as
``` c#
 (!string.IsNullOrEmpty(_tagHelperPrefix) &&
  (tagName.Length <= _tagHelperPrefix.Length ||
   !tagName.StartsWith(_tagHelperPrefix, StringComparison.OrdinalIgnoreCase))
```

dougbu commented on test/Microsoft.AspNet.Razor.Test/Generator/CSharpTagHelperRenderingTest.cs:
```diff
@@ -59,50 +30,64 @@ public static TheoryData TagHelperDescriptorFlowTestData
+                        "BasicTagHelpers.Prefixed",
+                        "BasicTagHelpers.Prefixed",
+                        BuildPAndInputTagHelperDescriptors(prefix: "THS"),
+                        BuildPAndInputTagHelperDescriptors(prefix: "THS"),
```
nit: create a `static` `PrefixedPAndInputTagHelperDescriptors` property right next to `DefaultPAndInputTagHelperDescriptors`

dougbu commented on test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs:
```diff
@@ -19,6 +19,282 @@ namespace Microsoft.AspNet.Razor.Test.TagHelpers
+            {
+                var factory = CreateDefaultSpanFactory();
+                var blockFactory = new BlockFactory(factory);
+                var availableDescriptors1 = new TagHelperDescriptor[]
```
pls give these more specific names.  bit easy to forget what "1" means by the end of this data set.  suggest `thPrefixedDescriptors`, `prefixPrefixedDescriptors`, `mythPrefixedDescriptors`

dougbu commented on test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs:
```diff
@@ -19,6 +19,282 @@ namespace Microsoft.AspNet.Razor.Test.TagHelpers
+                {
+                    new TagHelperDescriptor(
+                        prefix: "myth",
+                        tagName: "myth",
```
I wasn't specific enough in my `<myth/>` suggestion.  meant this should be a catch-all descriptor.  covers the case where Razor source uses a tag name that exactly matches the prefix of a catch-all descriptor.

dougbu replied:
(as-is, this descriptor would only match `<mythmyth ... />`)

dougbu commented:
:shipit: after reverting `TagHelperParseTreeRewriter` and making non-nit test changes; send out another iteration if I'm missing something about `TagHelperParseTreeRewriter`
